### ISSUE TITLE
Typescript errors that prevent projects including tsoa to be built

### ISSUE THREAD

**Issue by Jule-** [lukeautry/tsoa#1298]
## Sorting

- **I'm submitting a ...**

  - [x] bug report
  - [ ] feature request
  - [ ] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

No Typescript error when I try to run Typescript on a project including `tsoa` with Typescript v4.8.2.

## Current Behavior

```
node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:7:65 - error TS2344: Type 'T' does not satisfy the constraint 'Object | Function'.
  Type 'T' is not assignable to type 'Function'.

7 export declare function Middlewares<T>(...mws: Array<Middleware<T>>): ClassDecorator & MethodDecorator;
                                                                  ~

  node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:7:37
    7 export declare function Middlewares<T>(...mws: Array<Middleware<T>>): ClassDecorator & MethodDecorator;
                                          ~
    This type parameter might need an `extends Function` constraint.
  node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:7:37
    7 export declare function Middlewares<T>(...mws: Array<Middleware<T>>): ClassDecorator & MethodDecorator;
                                          ~
    This type parameter might need an `extends Object | Function` constraint.

node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:14:76 - error TS2344: Type 'T' does not satisfy the constraint 'Object | Function'.
  Type 'T' is not assignable to type 'Function'.

14 export declare function fetchMiddlewares<T>(target: any): Array<Middleware<T>>;
                                                                              ~

  node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:14:42
    14 export declare function fetchMiddlewares<T>(target: any): Array<Middleware<T>>;
                                                ~
    This type parameter might need an `extends Function` constraint.
  node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:14:42
    14 export declare function fetchMiddlewares<T>(target: any): Array<Middleware<T>>;
                                                ~
    This type parameter might need an `extends Object | Function` constraint.


Found 2 errors in the same file, starting at: node_modules/@tsoa/runtime/dist/decorators/middlewares.d.ts:7
```

## Possible Solution

Just forward the `extends` constraint of:
```ts
type Middleware<T extends Function | Object> = T;
```
to
```ts
export function Middlewares<T extends Function | Object>(...mws: Array<Middleware<T>>): ClassDecorator & MethodDecorator {
  //                          -------------------------
}

export function fetchMiddlewares<T extends Function | Object>(target: any): Array<Middleware<T>> {
  //                               -------------------------
}
```

## Steps to Reproduce

Open this repository and set your Typescript version to v4.8.2 in your IDE you should see these errors in the file https://github.com/lukeautry/tsoa/blob/master/packages/runtime/src/decorators/middlewares.ts.

## Context (Environment)

Version of the library: 4.1.1
Version of NodeJS: 16.14.2

**Comment by bryansamuel1** [comment]
This issue is happening for me as well. I am running through the steps found in Getting Started and hit the error when running the command `yarn run tsc --outDir build --experimentalDecorators` in the Building the routes file step

**Comment by elidaniel92** [comment]
**TL;DR** I solved my problem using typescript v4.7.4.

I also had a problem with Typescript v4.8.2. I started a tsoa project this week and on friday (26/08) everything broke. 

Environment:
- tsoa: 4.1.1;
- node: v16.14.0;
- npm, not using Yarn.

I tried to execute _npm run build_  but no route was generated in routes.ts,  all my controllers files was ignored.
```typescript
@Tags("Foo")
@Route("api/foo/bar")
export class FooController extends Controller {
```

I use typescript additional checks in my tsconfig.json
```json
"noUnusedLocals": true,
"noUnusedParameters": true,
```
So I get some errors on unused local variables and  on unused parameters in functions when I tried to execute _npm run build_:
```json
"build": "tsoa spec-and-routes && tsc",
```
```
src/routes.ts:4:117 - error TS6133: 'fetchMiddlewares' is declared but its value is never read.

4 import { Controller, ValidationService, FieldErrors, ValidateError, TsoaRoute, HttpStatusCodeLiteral, TsoaResponse, fetchMiddlewares } from 
'@tsoa/runtime';
                                                                                                                      ~~~~~~~~~~~~~~~~        

src/routes.ts:5:1 - error TS6133: 'expressAuthentication' is declared but its value is never read.

5 import { expressAuthentication } from './auth/tsoa-authentication-module';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/routes.ts:8:1 - error TS6133: 'iocContainer' is declared but its value is never read.

8 import { iocContainer } from './ioc/ioc';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/routes.ts:9:1 - error TS6192: All imports in import declaration are unused.

9 import { IocContainer, IocContainerFactory } from '@tsoa/runtime';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/routes.ts:10:1 - error TS6133: 'RequestHandler' is declared but its value is never read.

10 import type { RequestHandler } from 'express';
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/routes.ts:21:32 - error TS6133: 'app' is declared but its value is never read.

21 export function RegisterRoutes(app: express.Router) {
                                  ~~~

src/routes.ts:36:14 - error TS6133: 'promiseHandler' is declared but its value is never read.

36     function promiseHandler(controllerObj: any, promise: any, response: any, successStatus: any, next: any) {
                ~~~~~~~~~~~~~~

src/routes.ts:81:14 - error TS6133: 'getValidatedArgs' is declared but its value is never read.

81     function getValidatedArgs(args: any, request: any, response: any): any[] {
                ~~~~~~~~~~~~~~~~


Found 8 errors in the same file, starting at: src/routes.ts:4
```
I kept the typescript version fixed in 4.7.4 to resolve my problem.
```json
"typescript": "4.7.4"
````

**Comment by WoH** [comment]
https://github.com/lukeautry/tsoa/releases/tag/v5.0.0-alpha.0
